Sources/Testing/Traits/Trait.swift

```diff
@@ -81,6 +81,41 @@ public protocol SuiteTrait: Trait {
   var isRecursive: Bool { get }
 }
 
+/// A protocol extending ``Trait`` that offers an additional customization point
+/// for trait authors to execute code before and after each test function (if
+/// added to the traits of a test function), or before and after each test suite
+/// (if added to the traits of a test suite).
+public protocol CustomExecutionTrait: Trait {
+
+  /// Execute a function with the effects of this trait applied.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///   - testCase: The testCase under which `function` is being performed.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing the
+  ///   trait from running correctly.
+  ///
+  /// This function is called for each ``CustomExecutionTrait`` on a test suite
+  /// or test function and allows additional work to be performed before or
+  /// after the test runs.
+  /// It gets invoked once for each test, and then once for each test case.
+  ///
+  /// Issues recorded by this function are recorded against `test`. It is _not_
+  /// getting called for skipped tests.
```

Comment:
```suggestion
  /// called for skipped tests.
```

But also, these two sentences are not related. I think the new comment about skipped tests belongs in the previous paragraph?

Reply:
I removed that sentence since the Note below says pretty much the same thing.